device/chip2chip/chip2chip_base.c

```diff
@@ -0,0 +1,500 @@
+	while((CTC_In(rgstr_vptr.cmd) & CMD_READY_MASK) == 0){
+	//while((Xil_In64(C2C_CMD_ADDR) & CMD_READY_MASK) == 0){
+		if(time_cnt == 0)
+			return -1;
+		time_cnt--;
+	}
+	return 0;
+}
+
+int wait_wrData_ready(void){
+	int time_cnt = MAX_WR_DATA_READY_CNT;
+	/* command �����Ϳ��� 64-bit�� �о� �� �� write data ready ���� ����� �� write data ready ���� 0�̸� �ݺ�*/
+	while((CTC_In(rgstr_vptr.cmd) & WR_DATA_READY_MASK) == 0){
+	//while((Xil_In64(C2C_CMD_ADDR) & WR_DATA_READY_MASK) == 0){
+		if(time_cnt == 0)
+			return -1;
+		time_cnt--;
+	}
+	return 0;
+}
+
+int wait_writeData_req(u64* requested_tag){
+	u64 wr_er_status;
+	u64 wr_dataReq_ready;
+	u64 wr_dataReq_tag;
+	int time_cnt = MAX_WR_DATA_REQ_WAIT_CNT;
+
+	while(time_cnt-- != 0){
+		/* write & erase register�� ���� �����ִ� �� 64-bit �о��*/
+		wr_er_status = CTC_In(rgstr_vptr.wne_stat);
+		//wr_er_status = Xil_In64(C2C_WRITE_ERASE_STATUS_ADDR);
+		
+		/* write & erase register�� ���� �����ִ� �� 64-bit ���� write data request ready ��ȣ�� �����*/
+		wr_dataReq_ready = wr_er_status & WR_DATA_REQ_READY_MASK;
+		//wr_dataReq_ready = wr_er_status & WR_DATA_REQ_READY_MASK;
+		
+		/* write & erase register�� ���� �����ִ� �� 64-bit ���� write data request tag�� �����*/
+		wr_dataReq_tag = (wr_er_status & WR_DATA_REQ_TAG_MASK)>>WR_DATA_REQ_TAG_BIT;
+		
+		/* if write data request ready�� 0�� �ƴϸ� if�� ���� break*/
+		if(wr_dataReq_ready > 0){
+			/* write & erase register�� write data request valid bit�� 1�� ����*/
+			CTC_Out(rgstr_vptr.wne_stat, wr_er_status|WR_DATA_REQ_VALID);
+			//Xil_Out64(C2C_WRITE_ERASE_STATUS_ADDR, wr_er_status|WR_DATA_REQ_VALID);
+			
+			/* write & erase register�� write data request valid bit�� 0��� ����*/
+			CTC_Out(rgstr_vptr.wne_stat, wr_er_status&(~WR_DATA_REQ_VALID));
+			//Xil_Out64(C2C_WRITE_ERASE_STATUS_ADDR, wr_er_status&(~WR_DATA_REQ_VALID));
+			
+			/* write data request�� ���� tag�� �ٱ� �Լ�� ����ϱ� ��� ���� ����*/
+			*requested_tag = wr_dataReq_tag;
+			break;
+		}
+	}
+	if(time_cnt == 0)
+		return -1;
+	else
+		return 0;
+}
+
+int wait_flash_operation(u64 op, u64 tag, int* Qnumber, u64* ack,u64* ack_tag){
+	u64 readQ_ready;
+	u64 readQ_tag;
+	u64 wr_er_status;
+	u64 ack_ready;
+	int time_cnt;		//limit wait time
+
+	if(op == READ){
+		/*Ÿ�� ī��Ʈ ��� �̸� define �س�� read ������ �ִ� �ð� ����� ����Ʈ*/
+		time_cnt = MAX_READ_STATUS_WAIT_CNT;
+		/*loop�� �ݺ��� �� ���� Ÿ�� ī��Ʈ ��� ����*/
+		/*Ÿ�� ī��Ʈ ��==0 �̸� �ݺ��� break*/
+		while(time_cnt-- != 0){
+			/* read queue ready ��ȣ�� ����� (ready7, ready6, ready5, ..., ready0) */
+			readQ_ready = CTC_In(rgstr_vptr.read_stat) & READQ_READY_MASK;
+			//readQ_ready = Xil_In64(C2C_READ_STATUS_ADDR) & READQ_READY_MASK;
+			
+			/* read queue�� ����� data�� �ش��ϴ� tag�� ����� */
+			readQ_tag = CTC_In(rgstr_vptr.read_stat) & READQ_TAG_MASK;
+			//readQ_tag = Xil_In64(C2C_READ_STATUS_ADDR) & READQ_TAG_MASK;
+
+			if((readQ_ready&READ_DATA0_READY_VALUE)>0) // if read queue0�� ready ��ȣ�� 1�� ��

```

Comment:
For example,

> if you want to avoid stack growth, then please use the `static inline`.
> https://github.com/BlaCkinkGJ/Flash-Translation-Layer/blob/d74f061520ea7ce629074e4166929c06c9a14631/include/page.h#L101C1-L104

```c
const int read_ready_values[] = {
    READ_DATA0_READY_VALUE,
    READ_DATA1_READY_VALUE,
    READ_DATA2_READY_VALUE,
    READ_DATA3_READY_VALUE,
    READ_DATA4_READY_VALUE,
    READ_DATA5_READY_VALUE,
    READ_DATA6_READY_VALUE,
    READ_DATA7_READY_VALUE,
}

#define NUM_ITEMS(s) sizeof(s)/sizeof(s[0])

for (int i = 0; i < NUM_ITEMS(read_ready_values); i++) {
    if((readQ_read & read_ready_values[i]) > 0 ) { // you can make this part to a function.
        ...
    }
}
```